File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function averageOfSubtree(root: TreeNode | null): number {
    let count = 0;
    function dfs(node: TreeNode | null): [number, number] {
        if (!node) return [0, 0];
        const [ls, lc] = dfs(node.left);
        const [rs, rc] = dfs(node.right);
        const sum = ls + rs + node.val;
        const cnt = lc + rc + 1;
        if (Math.floor(sum / cnt) === node.val) count++;
        return [sum, cnt];
    }
    dfs(root);
    return count;
};

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn average_of_subtree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, count: &mut i32) -> (i64, i64) {
            if let Some(n) = node {
                let n = n.borrow();
                let (ls, lc) = dfs(&n.left, count);
                let (rs, rc) = dfs(&n.right, count);
                let sum = ls + rs + n.val as i64;
                let cnt = lc + rc + 1;
                if sum / cnt == n.val as i64 {   // integer division
                    *count += 1;
                }
                (sum, cnt)
            } else {
                (0, 0)
            }
        }
        let mut count = 0;
        dfs(&root, &mut count);
        count
    }
}